fix(nvcheatsheet): fix unicode highlighting and spacing of nvcheatsheet

kingavatar opened the pull request:
As the nvim_buf_add_highlight and other functions use byte indexing instead of character indexing, I have fixed the whitespace length to not include the extra bytes of unicode and also for highlighting I had to get the correct byte index for both text and space lines of cheatsheet using builtin functions.

Additional testing is required to see all cases are handled. I have checked for emojis and glyphs working as of now.

fixes #74

Before:
<img width="1278" alt="before" src="https://user-images.githubusercontent.com/31726036/230628914-8ba26dde-409d-403a-be24-7e6e6f769566.png">

After:
<img width="1278" alt="after" src="https://user-images.githubusercontent.com/31726036/230628980-108a58b9-05b2-4576-a0b9-577c3b29b460.png">

kingavatar committed:
fix(nvcheatsheet): fix unicode highlighting and spacing of nvcheatsheet

As the nvim_buf_add_highlight and other functions use byte indexing instead
of character indexing, I have fixed the whitespace length to not include the
extra bytes of unicode and also for highlighting I had to get the correct byte
index for both text and space lines of cheatsheet using builtin functions.

Additional testing is required to see all cases are handled.
I have checked for emojis and glyphs working as of now.

fixes #74

kingavatar committed:
fix(nvcheatsheet): handle title highlights as well

For Now this will handle spacing and highlights of title and space below as well.
But emoji in title is bit tricky to solve for now emojis or glyphs in title should not be allowed.

kingavatar commented:
@siduck  still heading with emojis are not handled. So For now we should not allow emojis or glyphs for headings.

siduck commented:
> @siduck  still heading with emojis are not handled. So For now we should not allow emojis or glyphs for headings.

You're right, adds a lil complexity  to the codebase too.

siduck commented:
@kingavatar wait, all we care about is nerd font icons, no real emojis!!!

kingavatar commented:
@siduck I meant unicodes. Both nerd font icons and emojis are unicodes so no difference. Both will work for cheatsheets but not for the heading titles.

<img width="962" alt="image" src="https://user-images.githubusercontent.com/31726036/230649221-ee1e147e-6016-4eba-afd9-74871c8d1f21.png">

It is hard to constrain heading title right now. In future PR I will try to fix this. For now emojis and nerd fonts can be used anywhere except for headings.

siduck commented:
oh, I'll wait till you fix it on headings then!

kingavatar committed:
fix(nvcheatsheet): fix heading nerd font as well

I have fixed nerd font icon for heading title as well. I have tried to include
emojis but yet they are causing the space highlighting to offset right side.
So emoji support yet.

kingavatar commented:
@siduck  The fix for nerd font was simple and I was wrong the emoji is causing to overshoot causing wrong alignment everywhere 😖. 

I can't seem to find easy fix and alternative is to check for all emoji left of character and account for the alignment which is a bad fix. 

So no emoji support for nvcheatsheet.

<img width="1274" alt="image" src="https://user-images.githubusercontent.com/31726036/230658681-6befe53a-4784-4b75-b3ed-394556508c20.png">

kingavatar commented:
@siduck  you can merge this now
